### PR TITLE
crypto/subscription_restart_on_fail

### PR DESCRIPTION
**Pull request by Binary-Mohanad**

### CODE REVIEW

File: lib/Net/Async/Blockchain/BTC.pm [hunk 1]
```diff
@@ -123,9 +125,18 @@ sub subscribe {
...
     die "Invalid or not implemented subscription" unless $subscription && $self->can($subscription);
+    my ($zmq_client_source, $zmq_client_socket) = $self->new_zmq_client->subscribe($subscription);
+
+    my $error_handler = sub {
+        my $error = shift;
+        $self->source->fail($error);
+        zmq_close($zmq_client_socket);
+    };
+
+## Please see file lib/Net/Async/Blockchain/BTC.pm.ERR
```
Comment (mwm-binary) [inline lib/Net/Async/Blockchain/BTC.pm]: This looks like some failed tidy-ing detritus.

File: lib/Net/Async/Blockchain/BTC.pm [hunk 2]
```diff
@@ -123,9 +125,18 @@ sub subscribe {
...
     Future->needs_all(
-        $self->new_zmq_client->subscribe($subscription)->map(async sub { await $self->$subscription(shift) })->ordered_futures->completed(),
-        $self->recursive_search());
+        $zmq_client_source->map(async sub { await $self->$subscription(shift) })->ordered_futures->completed(),
```
Comment (aminmarashi-binary) [inline lib/Net/Async/Blockchain/BTC.pm]: Seems a little bit suspicious, we handle failure here, but where do we handle success?

Comment (Binary-Mohanad) [reply lib/Net/Async/Blockchain/BTC.pm]: We process the received data from the source in the `map` which will do some operations on this data then emit the result to another source.

File: lib/Net/Async/Blockchain/BTC.pm [hunk 3]
```diff
@@ -123,9 +125,18 @@ sub subscribe {
...
+        $self->recursive_search())->on_fail($error_handler)->retain;
```
Comment (chengyang-binary) [inline lib/Net/Async/Blockchain/BTC.pm]: just to nit pick, shouldn't it be `on_fail(sub {...} )` ?

Comment (Binary-Mohanad) [reply lib/Net/Async/Blockchain/BTC.pm]: the `$error_handler` is an anonymous subroutine

Comment (chengyang-binary) [reply lib/Net/Async/Blockchain/BTC.pm]: ah ok. got it

**Pull request merged**